code/software/libs/src/sst_flash/sst.c

```diff
@@ -0,0 +1,139 @@
+/*
+ *------------------------------------------------------------
+ *                                  ___ ___ _   
+ *  ___ ___ ___ ___ ___       _____|  _| . | |_ 
+ * |  _| . |_ -|  _| . |     |     | . | . | '_|
+ * |_| |___|___|___|___|_____|_|_|_|___|___|_,_| 
+ *                     |_____|         tech demo
+ * ------------------------------------------------------------
+ * Copyright (c)2021 Ross Bamford
+ * See top-level LICENSE.md for licence information.
+ *
+ * SST39SF0x0 Flash EEPROM interface library
+ * ------------------------------------------------------------
+ */
+
+#include "sst_flash.h"
+
+#define SST_UNLOCK_A1       ((0x00005555<<1))
+#define SST_UNLOCK_D1       0xAA
+#define SST_UNLOCK_A2       ((0x00002AAA<<1))
+#define SST_UNLOCK_D2       0x55
+
+#define SST_SUBCMD_SECERASE 0x30
+#define SST_SUBCMD_CHPERASE 0x10
+
+static void write(uint32_t addr, uint8_t data) {
+    uint8_t *p = (uint8_t*)addr;
+    *p = data;
+}
+
+static void wait(uint32_t addr) {
+    volatile uint8_t *p = (uint8_t*)addr;
+
+    uint8_t last = (*p) & 0x40;
+    uint8_t this;
+
+    while ((this = ((*p) & 0x40)) != last) {
+        last = this;
+    }
+}
+
+static void unlock(uint32_t rom_base) {
```

Comment:
From the datasheet, it seems to me that the DQ6 toggling only happens for the Program and Erase operations, not after individual writes of the commands. I'd think that we only want to call `wait` in `sst_sector_erase`, `sst_sector_erase`, and `sst_write_byte`, but not in `unlock` or in `sst_command`.

Reply:
Interestingly, this was my read of the data sheet too, however in testing I found that, without the waits, it didn't work. I just tried it again and reliably get this:

```
Kernel received okay; Starting...


rosco_m68k Flash EEPROM identifier / write POC
**********************************************

EVEN ROM is  <Unknown>  [Manufacturer ???]
ODD  ROM is  <Unknown>  [Manufacturer ???]

Config area is 1 sector per ROM (2 total)
  Sector num : 0x7f
  Sector size: 4096 bytes
  Total size : 8192 bytes
  Base addr  : 0x00efe000

Old data: 0xaa55aa55 0xaa55aa55 0xaa55aa55...
New data: 0xaa55aa55 0xaa55aa55 0xaa55aa55...
```

_However_ , I now suspect that the call to write was just preventing an optimisation (which you mention in your other comment). Making the write value in `write` be volatile allows these `wait`s to be removed, while keeping everything working, so looks like this is a doubly-good shout, thanks! :)